libgadget/treewalk.c

```diff
@@ -1047,6 +1044,14 @@ ngb_treefind_threads(TreeWalkQueryBase * I,
             int i;
             int * suns = current->s.suns;
             for (i = 0; i < current->s.noccupied; i++) {
+                /* must be the correct type: compare the
+                 * current type for this subnode extracted
+                 * from the bitfield to the mask.*/
+                int type = (current->s.Types >> (3*i)) % 8;
```

Comment:
What about defining a per node type mask -- if the node has particle of that type then then the field is one. Then
the if below becomes
if (node->s.typemask & iter->mask)

Reply:
That doesn't really solve the problem here, because almost all nodes will have both gas and DM particles. We would still have to load all the particle data into memory for each child particle.